src/components/fs/FS.cpp

```diff
@@ -0,0 +1,194 @@
+#include <cstdlib>
+#include "drivers/SpiNorFlash.h"
+#include <littlefs/lfs.h>
+#include <lvgl/lvgl.h>
+
+using namespace Pinetime::Controllers;
+
+
+static constexpr size_t BLOCK_SIZE_BYTES = 4096;
+
+/*
+*   Interface between littlefs and SpiNorFlash
+*/
+
+namespace {
+    static int read(const struct lfs_config *c, lfs_block_t block,
+                    lfs_off_t off, void *buffer, lfs_size_t size) {
+        Pinetime::Controllers::FS& lfs = *(reinterpret_cast<Pinetime::Controllers::FS*>(c->context));
+        const size_t address = lfs.mStartAddress + (block * BLOCK_SIZE_BYTES) + off;
+        lfs.mDriver.Read(address, (uint8_t*)buffer, size);
+
+        return 0u;
+    }
+
+    static int prog(const struct lfs_config *c, lfs_block_t block,
+            lfs_off_t off, const void *buffer, lfs_size_t size) {
+        Pinetime::Controllers::FS& lfs = *(reinterpret_cast<Pinetime::Controllers::FS*>(c->context));
+        const size_t address = lfs.mStartAddress + (block * BLOCK_SIZE_BYTES) + off;
+        lfs.mDriver.Write(address, (uint8_t*)buffer, size);
+        return lfs.mDriver.ProgramFailed() ? -1u : 0u;
+    }
+
+    static int erase(const struct lfs_config *c, lfs_block_t block) {
+        Pinetime::Controllers::FS& lfs = *(reinterpret_cast<Pinetime::Controllers::FS*>(c->context));
+        const size_t address = lfs.mStartAddress + (block * BLOCK_SIZE_BYTES);
+        lfs.mDriver.SectorErase(address);
+        return lfs.mDriver.EraseFailed() ? -1u : 0u;
+    }
+
+    static int sync(const struct lfs_config *c) {
+        // no hardware caching used
+        return 0u;
+    }
+}
+
+const static struct lfs_config baseLfsConfig = {
+    .read = read,
+    .prog = prog,
+    .erase = erase,
+    .sync = sync,
+
+    .read_size = 16,
+    .prog_size = 8,
+    .block_size = BLOCK_SIZE_BYTES,
+    .block_cycles = 1000u,
+
+    .cache_size = 16,
+    .lookahead_size = 16,
+
+    .name_max = 50,
+    .attr_max = 50
+
+};
+
+constexpr struct lfs_config createLfsConfig(Pinetime::Controllers::FS& fs, const size_t totalSizeBytes) {
+    struct lfs_config config = baseLfsConfig;
+    config.context = &fs;
+    config.block_count = totalSizeBytes / BLOCK_SIZE_BYTES;
+    return config;
+}
+
+FS::FS(Pinetime::Drivers::SpiNorFlash& driver) :
+    mDriver{driver},
+    mLfsConfig{createLfsConfig(*this, mSize)} { }
+
+
+void FS::Init() {
+
+    // try mount
+    int err = lfs_mount(&mLfs, &mLfsConfig);
+
+    // reformat if we can't mount the filesystem
+    // this should only happen on the first boot
+    if (err != 0) {
+        lfs_format(&mLfs, &mLfsConfig);
+        err = lfs_mount(&mLfs, &mLfsConfig);
+        if (err != 0) {
+            return;
+        }
+    }
+
+#ifndef PINETIME_IS_RECOVERY    
+    VerifyResource();    
+    LVGLFileSystemInit();
+#endif
+
+}
+
+void FS::VerifyResource() {
+    // validate the resource metadata
+    resourcesValid = true;
+}
+
+void FS::FileOpen(lfs_file_t* file_p, const char* fileName, const int flags) {
```

Comment:
Shouldn't these functions return the error code that the lfs functions might return?